src/wallet/wallet.cpp

```diff
@@ -1118,16 +1115,12 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFlushOnClose)
     bool fUpdated = false;
     if (!fInsertedNew)
     {
-        // Merge
-        if (!wtxIn.hashUnset() && wtxIn.hashBlock != wtx.hashBlock)
-        {
-            wtx.hashBlock = wtxIn.hashBlock;
-            fUpdated = true;
-        }
-        // If no longer abandoned, update
-        if (wtxIn.hashBlock.IsNull() && wtx.isAbandoned())
-        {
+        assert(!wtxIn.isAbandoned()); // Incoming transactions should never have special abandoned block hash
```

Comment:
After reading again about how to use abandontransaction, I think you're right, tx might be not in a block or not in your mempool so you can flag is as abandoned. But after this, it can still being included as part of other peers mempools. Removed it, was suggested here initially https://github.com/bitcoin/bitcoin/pull/15931#discussion_r295463739